soroban-env-common/src/raw_val.rs

```diff
@@ -183,6 +183,60 @@ impl Tag {
         // The `test_tag_from_u8` test should ensure this cast is correct.
         unsafe { ::core::mem::transmute(tag) }
     }
+
+    /// Get the ScValType of the XDR type that corresponds to this tag.
+    ///
+    /// For use in the `Host::obj_cmp` comparison function so that comparison
+    /// based on tags can be done identically to the `ScVal` type.
+    ///
+    /// Returns `None` for `Tag::Bad`.
+    ///
+    /// # Panics
+    ///
+    /// If `self` is any of the three marker tags,
+    /// `SmallCodeUpperBound`, `ObjectCodeLowerBound`, `ObjectCodeUpperBound`.
+    /// These should not be constructed under code paths that might use this function.
+    #[inline(always)]
+    pub const fn get_scval_type(&self) -> Option<ScValType> {
+        match *self {
+            Tag::False => Some(ScValType::Bool),
+            Tag::True => Some(ScValType::Bool),
+            Tag::Void => Some(ScValType::Void),
+            Tag::Status => Some(ScValType::Status),
+            Tag::U32Val => Some(ScValType::U32),
+            Tag::I32Val => Some(ScValType::I32),
+            Tag::U64Small => Some(ScValType::U64),
+            Tag::I64Small => Some(ScValType::I64),
+            Tag::TimepointSmall => Some(ScValType::Timepoint),
+            Tag::DurationSmall => Some(ScValType::Duration),
+            Tag::U128Small => Some(ScValType::U128),
+            Tag::I128Small => Some(ScValType::I128),
+            Tag::U256Small => Some(ScValType::U256),
+            Tag::I256Small => Some(ScValType::I256),
+            Tag::SymbolSmall => Some(ScValType::Symbol),
+            Tag::LedgerKeyContractExecutable => Some(ScValType::LedgerKeyContractExecutable),
+            Tag::SmallCodeUpperBound => panic!(),
```

Comment:
Since this code is called in the host and is under the control of the guest, I think it has to be panic-free. Unless I'm misunderstanding how that's being ruled out.

Reply:
The three panic cases are the "marker" tags, which I think can't be constructed under guest-controlled code paths. The only calls to `get_scval_type` are in `obj_cmp` here:

```rust
            None => {
                let atype = a.get_tag().get_scval_type();
                let btype = b.get_tag().get_scval_type();
                if atype == btype {
                    // This shouldn't have happened, but if it does there's a logic error.
                    return Err(
                        self.err_general("equal-tagged values rejected by small-value obj_cmp")
                    );
                }
                atype.cmp(&btype)
            }
```

Called on the result of `get_tag` which returns `Tag::Bad` for the cases that might otherwise return the three marker tags.

So I think there is no way for a guest to call this function and trigger a panic.

I debated this myself though and can appreciate an argument for being defensive here and returning `None`.

Reply:
I have removed the panics.

Reply:
Marker tags can definitely be constructed by guest code -- guest code can construct a `RawVal` with any bit pattern it likes -- but I see what you mean now that the panicking calls should be guarded by `get_tag()`. But yeah, in general I'm aiming for as few such implicit "this panic is safe because..." latent assumptions in the code; it'll be too easy for someone to break such an assumption with a future refactor. Thanks!